### PR TITLE
Add higher-order createType functions

### PR DESCRIPTION
**Pull request by xaviervia**
- [x] Setoid
- [x] Functor
- [x] Semigroup
- [x] Monoid
- [x] Foldable

### CODE REVIEW

**Comment by xaviervia** [comment]
Mysteriously enough, [type checking of a Functor](https://flowtype.org/try/#0FAUwHgDg9gTgLgAjgTwiBAVVIA8BBAGgQCEA+BAXgQG9gEEByAASZTQYC4FC7GWA3AIYAbAK4hOJArwCWAOwDOaAMZwAFAEouAZTgx5Ac2n0ZCtWC4ARQXEE4safETJFBc5KS0IARlCjCQN2AAX2BQSFhENnQAMVE5VVgnEnIqB1xCFIQAMhpeAFtBCDUAMy41QS5iDUpyaq44hLgkzLIQ4GiEAFki-FSEcy48GopyUvLK7hHyYdruMPBoeCRsBGtbZLJKPPpmVmxJN2RjBCgygYsSafnQgHoAKjp7hABxEDkQfWVMVYBhGECzRgADoEABJEoIUQKEAAEwQAHcZHAABYIGLCKAIhgKcFyEqfd7KdDvAzyEBEBGwADWuIU+T8qOEyCeSCgCH4nxkJWQSBRNgQAAMzoKfJiEbjRBAEPIkDJ8ugSrA+egETAoHIDCs0ApgU9bsBlBqFIhlACbCB0ts1HJBAquEcvEc5mpaLsWNFJLaFSczlwwC63fQ+Pt2FxvRTeO6mEIxBJ-ScTIoVHBytdBQASagR4JqLNgYIaQWJmUKLi80YIZAAbT2noAupQKFQI7xCyENGEjYpECV4okYNbChAuD0IDgjp4XWbAZbsOUIw73NdHS7FwhV5W1Gpotcg92TadIVQA1ug-RgZfosCzuYNCXh1xj+Rb8O76UNJ36O3eAC4KIYDkHZg0vYFohLM42w7NQZwtdIbTtEBPzCTowVhd44GRZBGgHDBthwoEcAYNCMKwgjYAYIg5FEfJvE+UhDWNRASLkTCUHImArBsOwWLY7D+yBDAqJouiYFSXg+yaWBzDmZ8EBKO8NBg804DnNAlOI9DWLIgSKM7Riew3bZeJ0qSQVvAAmDRgTfU9yADABqBAAGYNGrAAiAQRHEdz6wMw9vGMrS+I4m8FIAVjczyY28kBfLCQQmyobwUNWXR9E1Di8KoDiiPSwwOMohATQygwGIPRB8sy3TOLWbicCqgwsqIErDHE+hJIHGTKzkhSwE-ZTZ3SDTGsK-SKoQb4qFGmqwrUBgLIYazbLmRyXKiry43iib4WmvQCtm28GHCpaPM2ny-KAA) expecting map to go back to the same type works, but [the same type checking for a SemiGroup](https://flowtype.org/try/#0PQKgBAAgZgNg9gdzCYAoVBTAHgBzgJwBcxCBPHDMAFXIwB4BBAGjACEA+MAXjAG9UwYAOQQIZCkIBcYZgOGiAbgEMYAVwxS2TVAF902PERK0wAESWEljFh2585IsbU1KAdqW2C4UaQAos0qwAlNycDLr6uATE4pQAyhgAtgCWAOL4cKo41mycPDQUObYAZPaCAMZwruUW-tIJKemZ2cy5QfVJaRlZRewRsWAAwlU1hIx5YHUyIVycUwwzYeiVrgDOxKudTVl2viuj0sPVFnRupOyLk65KiRjSZ5f8FVXrYN52WKGTT4LyThLSa63Ty-RzKNQaaRYEHPY6EaSkL7ePYjWpYIK+UgAbTBKnUQgAukEgnIdCS5PgMIRVPhXGVQaJYpogRgYW8oKSIqgBnFCPhkq4AOYNZKC7o4Kh2EXbbJCXn8oUisXNIQsdYKwV9FaveUC4WdZVZaTmSx0XWKg3iqhqvl6vKoACQm0a4v8X0RszAAAMACS8LA6AB0ftIOi9GLltotKUNOCE5O1xCUUqj+pj4sDyKESnjqETYAARimNUqM1mC7nuSYAJKuQil5qSnjS8V0IS1+uWlUsVyqRIFjD4LUvYgdhtGswWKxjrtZa1gXv9wf2p1bV2fT0ezhYADUpAjM-TKoTI7A5Tsh9FZagvgAjCT8wATC918c4TM3gBMQSAA) blows up.

File: src/type.js [hunk 1]
```diff
@@ -11,8 +11,8 @@ export type TypeChecker<A> = (a: A) => A
 export type Type<A, B> = {
   '@@type': A,
   '@@value': B,
-  inspect(): String,
-  is(x: Data<Type<A, B>, any>): boolean
+  inspect: () => string,
+  is: (x: Data<Type<A, B>, any>) => boolean,
```
Comment (xaviervia) [inline src/type.js]: Why was this change needed? Out of curiosity.

Comment (ostera) [inline src/type.js]: We seem to be defining function signatures as `type Name = (a: A) => A` everywhere, but in the compound type definitions we were using the `name(a: A): A` syntax instead.

This makes it the same everywhere.
